packages/api/src/middlewares/withTelemetry.ts

```diff
@@ -1,43 +1,98 @@
 
+type APICounter = {
+  scope?: string;
+};
+
+type RequestsTotalCounter = APICounter & {
+  endpoint: string;
+  status_code: string;
+  method: string;
+};
+
+const apiRequestsTotalCounter = meter.createCounter<RequestsTotalCounter>(
+  "blobscan_api_requests_total",
+  {
+    valueType: api.ValueType.INT,
+    description:
+      "Number of requests made to Blobscan API by endpoint, status_code and method",
+  }
+);
+
+const apiRequestsDurationMsHistogram = meter.createHistogram<APICounter>(
+  "blobscan_api_requests_duration_ms",
+  {
+    description: "Duration of all the requests made to Blobscan API",
+    valueType: api.ValueType.INT,
+    unit: "ms",
+  }
+);
+
+function getEndpoint(url: string) {
+  return url.split("?")[0] ?? "";
+}
+
 function getProcedureFromUrl(url: string) {
-  const procedure = url
+  const procedure = getEndpoint(url)
     .split("/")
     .find((_, i, urlSplitted) => {
       const prevPath = urlSplitted[i - 1];
 
       return prevPath === "trpc";
-    })
-    // remove query params
-    ?.split("?")[0];
+    });
 
   return procedure;
 }
-export const withTelemetry = t.middleware(({ ctx: { req }, next }) => {
-  const procedureName = req?.url ? getProcedureFromUrl(req.url) : "unknown";
-  const spanOptions: api.SpanOptions = {};
-  let spanName: string;
-
-  if (procedureName) {
-    spanName = "trpc";
-    spanOptions.attributes = {
-      procedure: procedureName,
-    };
-  } else {
-    spanName = "trpc_open_api";
-    spanOptions.attributes = {
-      path: req.url ?? "unknown-path",
-    };
-  }
 
-  return tracer.startActiveSpan(spanName, spanOptions, async (span) => {
-    const result = await next();
+export const withTelemetry = t.middleware(
+  async ({ ctx: { req, res, scope }, next }) => {
+    const endpoint = req?.url ? getEndpoint(req.url) : "unknown-endpoint";
+    const procedureName = req?.url ? getProcedureFromUrl(req.url) : "unknown";
+    const spanOptions: api.SpanOptions = {};
+    let spanName: string;
+
+    if (procedureName) {
+      spanName = "trpc";
+      spanOptions.attributes = {
+        procedure: procedureName,
+      };
+    } else {
+      spanName = "trpc_open_api";
+      spanOptions.attributes = {
+        path: endpoint,
+      };
+    }
+
+    const requestStart = performance.now();
+    try {
+      const result = await tracer.startActiveSpan(
+        spanName,
+        spanOptions,
+        async (span) => {
+          const result = await next();
 
-    span.end();
+          span.end();
 
-    return result;
-  });
-});
+          return result;
+        }
+      );
+
+      return result;
+    } finally {
+      const requestEnd = performance.now();
+
+      apiRequestsDurationMsHistogram.record(requestEnd - requestStart, {
+        scope,
+      });
+      apiRequestsTotalCounter.add(1, {
+        scope,
+        status_code: res?.statusCode?.toString() ?? "unknown-status-code",
+        endpoint,
+        method: req?.method ?? "unknown-method",
```

Comment:
In which case does a http request not have method?

Reply:
There isn't a specific case. This check is in place to satisfy TypeScript typing.

In Node.js, both requests and responses are instances of the `IncomingMessage` class because they share many fields and methods. As a result, certain fields are marked as optional because they aren't applicable to both types (e.g. `method` on server responses).

Reply:
I'll use an empty string instead of `unknown-status-code` to make it less confusing.